amazon/ionbenchmark/ion_benchmark_cli.py

```diff
@@ -84,52 +100,169 @@
+def generate_simpleion_read_test_code(file, memory_profiling, io_type, iterator=False, single_value=False,
                                       emit_bare_values=False):
-    if not memory_profiling:
-        if not iterator:
-            def test_func():
-                with open(file, "br") as fp:
-                    data = ion.load(fp, single_value=single_value, emit_bare_values=emit_bare_values)
-                return data
+    if io_type == Io_type.BUFFER.value:
+        with open(file, "br") as fp:
+            benchmark_data = fp.read()
+        if not memory_profiling:
+            if not iterator:
+                def test_func():
+                    data = ion.loads(benchmark_data, single_value=single_value, emit_bare_values=emit_bare_values)
+                    return data
+            else:
+                def test_func():
+                    it = ion.loads(benchmark_data, single_value=single_value, emit_bare_values=emit_bare_values,
+                                   parse_eagerly=False)
+                    while True:
+                        try:
+                            next(it)
+                        except StopIteration:
+                            break
+                    return it
         else:
-            def test_func():
-                with open(file, "br") as fp:
-                    it = ion.load(fp, single_value=single_value, emit_bare_values=emit_bare_values,
-                                  parse_eagerly=False)
+            if not iterator:
+                def test_func():
+                    tracemalloc.start()
+                    data = ion.loads(benchmark_data, single_value=single_value, emit_bare_values=emit_bare_values)
+                    global read_memory_usage_peak
+                    read_memory_usage_peak = tracemalloc.get_traced_memory()[1] / BYTES_TO_MB
+                    tracemalloc.stop()
+                    return data
+            else:
+                def test_func():
+                    tracemalloc.start()
+                    it = ion.loads(benchmark_data, single_value=single_value, emit_bare_values=emit_bare_values,
+                                   parse_eagerly=False)
                     while True:
                         try:
                             next(it)
                         except StopIteration:
                             break
-                return it
+                    global read_memory_usage_peak
+                    read_memory_usage_peak = tracemalloc.get_traced_memory()[1] / BYTES_TO_MB
+                    tracemalloc.stop()
+                    return it
     else:
-        if not iterator:
+        if not memory_profiling:
+            if not iterator:
+                def test_func():
+                    with open(file, "br") as fp:
+                        data = ion.load(fp, single_value=single_value, emit_bare_values=emit_bare_values)
+                    return data
```

Comment:
> if that's possible to merge the similar logic in two different conditions into one reusable function?

Good question. We may not want to merge them into one function now but it's possible to refactor some logics if `test_func()` becomes complex later.

This is because the whole functionality of the `test_func` will be benchmarked and they (`load` and `loads`) are different APIs with a similar signature.

Reply:
> if that's possible to merge the similar logic in two different conditions into one reusable function?

Good question. We may not want to merge them into one function now but it's possible to refactor some logics if `test_func()` becomes complex later.

This is because the whole functionality of the `test_func` will be benchmarked and they (`load` and `loads`) are different APIs with a similar signature.